### PR TITLE
MAINT: special: add function `js_div` used by `spatial.distance.jensenshannon`

### LINKED ISSUE #19436: BUG: Jensen-Shannon Distance returns inf
### Describe your issue.

The ouput of the Jensen-Shannon Distance by definition is confined to the interval [0, 1]. I found an edge case where it returns inf. 

### Reproducing Code Example

```python
from scipy.spatial.distance import jensenshannon

p = [0, 1]
q = [5e-324, 1]

x = jensenshannon(p, q)
print(x)
```

### Error message

```shell
I expect a return value of about 1 but get inf.
```

### SciPy/NumPy/Python version and system information

```shell
Python 3.11.5 | packaged by Anaconda, Inc. | (main, Sep 11 2023, 13:26:23) [MSC v.1916 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.16.1 -- An enhanced Interactive Python. Type '?' for help.
PyDev console: using IPython 8.16.1
Python 3.11.5 | packaged by Anaconda, Inc. | (main, Sep 11 2023, 13:26:23) [MSC v.1916 64 bit (AMD64)] on win32
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info); scipy.show_config()
1.11.3 1.26.1 sys.version_info(major=3, minor=11, micro=5, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /c/opt/64/include
    lib directory: /c/opt/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= SKYLAKEX MAX_THREADS=2
    pc file directory: c:/opt/64/lib/pkgconfig
    version: 0.3.21.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /c/opt/64/include
    lib directory: /c/opt/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= SKYLAKEX MAX_THREADS=2
    pc file directory: c:/opt/64/lib/pkgconfig
    version: 0.3.21.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.11.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.36
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  pythran:
    include directory: C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-dic8cvan\overlay\Lib\site-packages/pythran
    version: 0.14.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
Python Information:
  path: C:\Users\runneradmin\AppData\Local\Temp\cibw-run-_etjukwv\cp311-win_amd64\build\venv\Scripts\python.exe
  version: '3.11'
```

### PR DESCRIPTION
**Pull request by fancidev**
#### Reference issue
Closes gh-19436, gh-20083.

#### What does this implement/fix?
Adds a function `special.js_div(x,y)` to compute the Jensen-Shannon divergence accurately. In particular, it ensures the result is non-negative. The referenced issues are both caused by negative distance due to numerical (cancellation) error.

`x` and `y` must be finite and nonnegative.

#### Additional information
This PR is intended to supersede PRs gh-19438 and gh-20414. The former one only addresses subnormal input (which does not help gh-20083), and the latter appears to cause [reduced accuracy](https://github.com/scipy/scipy/pull/20414#issuecomment-2044225890).

Test cases for the linked issues are added. Both the Python path (via `jensenshannon`) and the C path (via `cdist`) are covered.

### CODE REVIEW

**Comment by fancidev** [comment]
**Algorithm description**

Let $p_1, \cdots, p_n$ and $q_1, \cdots, q_n$ be probabilities such that $0 \le p_i,q_i\le 1$ and $p_1+\cdots+p_n = q_1+\cdots+q_n = 1$. The Jensen-Shannon distance is computed by

$$
\begin{align*}
\textrm{JSD} &= \sqrt{ \frac{1}{2} \left( \sum_{i=1}^n p_i\ln \frac{p_i}{(p_i+q_i)/2} + \sum_{i=1}^n q_i\ln \frac{q_i}{(p_i+q_i)/2} \right)} \\
&= \sqrt{ \sum_{i=1}^n h(p_i,q_i) }
\end{align*}
$$

where

$$
\begin{align*}
h(a,b) &:= \frac{1}{2} \left( a \ln a +b \ln b\right) -\frac{a+b}{2}\ln\frac{a+b}{2} \\
&= \frac{1}{2}\left[ f(a)+f(b)-2f\left(\frac{a+b}{2}\right) \right]
\end{align*}
$$

and $f(x) \equiv x\ln x$. Since $f(x)$ is convex, $h(a,b)$ is nonnegative by Jensen's inequality. However, numerical error may cause the evaluated result to be negative, as shown by the linked issues. This is related to the fact that $h(a,b)$ is proportional to the second order central  difference of $f(x)$ at $x=(a+b)/2$.

This PR adds the `js_div` function to evaluate $h(a,b)$ accurately and in particular to ensure the result is nonnegative. Let

$$
c \equiv \frac{a+b}{2}, d \equiv \frac{b-a}{2}, t \equiv \frac{d}{c} .
$$

Note that $-1 \le t \le 1$. Rewrite $h(a,b)$ as

$$
\begin{align*}
h(a,b) &= \frac{1}{2} \left( a\ln\frac{a}{c} + b\ln\frac{b}{c} \right)\\
&= \frac{c}{2} \left( \frac{a}{c}\ln\frac{a}{c} + \frac{b}{c}\ln\frac{b}{c} \right)\\
&=  \frac{c}{2} \left[ \left(1-\frac{d}{c}\right) \ln\left(1-\frac{d}{c}\right) + \left(1+\frac{d}{c}\right)\ln\left(1+\frac{d}{c}\right)\right)\\
&= \frac{c}{2} \left[ (1-t)\ln(1-t) + (1+t)\ln(1+t)\right] & \textrm{(1)}\\
&= \frac{c}{2} \left[ t(\ln(1+t)-\ln(1-t)) + (\ln(1+t)+\ln(1-t))\right]\\
&=c \left[ t \mathop{\mathrm{atanh}}(t) + \frac{1}{2} \ln(1-t^2)\right] & \textrm{(2)}\\
\end{align*}
$$

We now compare the cancellation error of formula (1) and (2), _assuming t is known exactly_. The Taylor series for the two terms of formula (1) are

$$
\begin{align*}
(1-t)\ln(1-t) &= -t+\frac{t^2}{2}+\frac{t^3}{2\cdot 3}+\frac{t^4}{3\cdot 4}+\frac{t^5}{4 \cdot 5} +\cdots\\
(1+t)\ln(1+t) &= +t+\frac{t^2}{2}-\frac{t^3}{2\cdot 3}+\frac{t^4}{3\cdot 4}-\frac{t^5}{ 4\cdot 5} +\cdots\\
\end{align*}
$$

When the two terms are summed, the leading term, as well as all odd powers of $t$, are cancelled. The cancellation error is significant if $t$ is close to zero. On the other hand, both series converge when $t=\pm 1$.

In comparison, the Taylor series of the two terms of formula (2) are

$$
\begin{align*}
t \mathop{\mathrm{atanh}}(t) &= +\frac{ t^2}{1} + \frac{t^4}{3} + \frac{t^6}{5} + \cdots \\
\frac{1}{2} \ln(1-t^2) &= -\frac{t^2}{2} - \frac{t^4}{4} - \frac{t^6}{6} - \cdots \\
\end{align*}
$$

Each term in the first series is larger in magnitude than the corresponding term in the second series. In particular, when $t$ is small, $t^2$ dominates, and the first series is about twice the magnitude of the second term, so cancellation error may not be significant. On the other hand, both series diverge at $t=\pm 1$.

Next, we compare the cancellation error of formula (1) and (2) empirically, again _assuming t is known exactly_. All terms but the final addition are correctly rounded (using `mpmath`), so that only cancellation error is exhibited.

<details>

<summary>
Script to compute cancellation error.
</summary>

```python
import numpy as np
import matplotlib.pyplot as plt
from mpmath import mp
mp.dps = 1000

def g0(t):
    t = mp.mpf(t)
    return float( (1-t)*mp.log(1-t) + (1+t)*mp.log(1+t) )

def g1(t):
    t = mp.mpf(t)
    return float((1-t)*mp.log(1-t)) + float((1+t)*mp.log(1+t))

def g2(t):
    t = mp.mpf(t)
    return float(mp.log1p(-t**2)) + float(2*t*mp.atanh(t))

ts = np.linspace(0.1, np.nextafter(1,0), 1000)
v0s = np.array([g0(t) for t in ts])
v1s = np.array([g1(t) for t in ts])
v2s = np.array([g2(t) for t in ts])

plt.plot(ts, v1s/v0s-1, ls='none', marker='x', label=r'Formula 1: $(1-t)\ln(1-t)+(1+t)\ln(1+t)$')
plt.plot(ts, v2s/v0s-1, ls='none', marker='.', label=r'Formula 2: $2t \text{ atanh}(t)+\ln(1-t^2)$')
plt.legend()
plt.grid()
plt.xlabel('$t$')
plt.ylabel('Relative Error')
plt.show()
```

</details>

image

The chart shows that for $t \ge 0.5$, formula (1) has slightly smaller error. For $t < 0.5$, the error of formula (1) grows as $t$ decreases. In fact, the graph starts from $t = 0.1$ because the error of formula (1) "explodes" as $t$ goes to zero.

The above comparison omits two other sources of error. First, $t$ is computed from $a$ and $b$, which has lower precision when $t > 0.5$. This favors formula (1) because it doesn't require computing $t$. Second, `log`, `log1p`, and `atanh` may not be correctly rounded in an actual implementation. We can't do much about this other than hoping they're not too off.

To summarise, given $0 \le a,b < +\infty$, `_js_div` evaluates $h(a,b)$ as follows:

$$
\begin{cases}
\frac{\ln 2}{2} (a+b) & \textrm{if } ab=0 \\
c \left[ t\mathop{\mathrm{atanh}}(t)+\frac{1}{2} \ln(1-t^2) \right] & \textrm{if } ab > 0, |t| \le 0.5 \\
\frac{1}{2} [ a\ln(a/c)+ b\ln(b/c) ] & \textrm{if } ab > 0, |t| > 0.5
\end{cases}
$$

where $c \equiv (a+b)/2$, $t \equiv (a-b)/(a+b)$. Provided $c$ and $t$ are evaluated without premature overflow, the result is finite for any finite input.

Special values are handled as follows. If both inputs are zero, return `+0.0`. If any input is negative or infinite, return `+inf` as a convex extension. If any input is nan, return `nan`.

**Comment by fancidev** [comment]
I have two questions regarding the code organization.

1/ Are you ok to implement the special function in the way of this PR? Or do you prefer to implement the special function along with kl_div and friends, i.e. in [_convex_analysis.pxd](https://github.com/scipy/scipy/blob/1ba35d1c4f66d82e0d395cf086e2212c06d8852d/scipy/special/_convex_analysis.pxd#L13)?

2/ Do you have any concern that the code is “duplicated” in two places? One used by the Python `jensenshannon` function, and the other used by `cdist`/`pdist`? This appears to be a common issue of similar functions. What’s worse, the duplicates are not identical but translated between C and C++ (or Cython)!

**Comment by fancidev** [comment]
I’m converting this PR to Draft for the moment as I’m working to consolidate the C++ and Python implementations of `spatial` distance metrics. Once that is done, the `jensenshannon` fix would be conveniently applied (to the C++ implementation) and there would be no need to add a special function for that purpose.

**Comment by tylerjereddy** [comment]
> I’m converting this PR to Draft for the moment as I’m working to consolidate

I'm going to use that as an excuse to bump the milestone for now, since we're getting closer to branching and this probably need a bit of thoughtful review even when it is ready to go.

**Comment by ev-br** [comment]
The PR is still a draft, has accumulated conflicts, and it's generally unclear if it's sill WIP. 
Given that many kernels from scipy.special now live in https://github.com/scipy/xsf, I also wonder if this implementation should also live in xsf.
Since it's not mergeable today, I'll bump the milestone again.  Would be great if you could revisit this @fancidev !

**Comment by ev-br** [comment]
We bumped the milestone once, and now the next release is approaching, https://discuss.scientific-python.org/t/proposed-release-schedule-for-scipy-1-17-0/2146. Would be great if you've bandwidth to return to it @fancidev

**Comment by lucascolley** [comment]
removing the milestone as this is still in draft

**Comment by fancidev** [comment]
@ev-br Thanks for the heads up. Indeed I think it would be a good fit for xsf. The computational logic for JS divergence is pretty much done; I was stuck with the integration into the distance metrics which did not "link" xsf. Let me revisit before year end!

**Comment by fancidev** [comment]
I'm going down a less-ambitious route -- instead of adding `js_div` to `xsf`, I'm adding it to [_convex_analysis.pxd](https://github.com/scipy/scipy/blob/063e2590aa44e9ffbc8689b4ded21657a8af4708/scipy/special/_convex_analysis.pxd#L4), joining its friends `entr`, `rel_entr`, `kl_div`, `huber` and `pseudo_huber`. I think it would be more efficient to move all these functions into `xsf` in a separate PR (if we decide to do so) since they are closely related.

**Comment by steppi** [comment]
> I'm going down a less-ambitious route -- instead of adding `js_div` to `xsf`, I'm adding it to [_convex_analysis.pxd](https://github.com/scipy/scipy/blob/063e2590aa44e9ffbc8689b4ded21657a8af4708/scipy/special/_convex_analysis.pxd#L4), joining its friends `entr`, `rel_entr`, `kl_div`, `huber` and `pseudo_huber`. I think it would be more efficient to move all these functions into `xsf` in a separate PR (if we decide to do so) since they are closely related.

That sounds reasonable. Good to see you back @fancidev!

**Comment by fancidev** [comment]
> > I'm going down a less-ambitious route -- instead of adding `js_div` to `xsf`, I'm adding it to [_convex_analysis.pxd](https://github.com/scipy/scipy/blob/063e2590aa44e9ffbc8689b4ded21657a8af4708/scipy/special/_convex_analysis.pxd#L4), joining its friends `entr`, `rel_entr`, `kl_div`, `huber` and `pseudo_huber`. I think it would be more efficient to move all these functions into `xsf` in a separate PR (if we decide to do so) since they are closely related.
> 
> That sounds reasonable. Good to see you back @fancidev!

Thanks Steppi ! The CI errors are now fixed, and I think it's ready for review!

**Comment by fbourgey** [comment]
Can you please first open a PR over at https://github.com/scipy/xsf?

**Comment by steppi** [comment]
> Can you please first open a PR over at https://github.com/scipy/xsf?

This PR predates `xsf` by quite a bit and as far as I know, fancidev is not active at the moment. In addition, when he was active, he expressed an interest in adding this to `xsf`, see https://github.com/scipy/xsf/issues/1#issuecomment-2388163621.
